add migration for home page setup

umarmughal824 opened the pull request:
#### What are the relevant tickets?
Closes #596 
https://trello.com/c/YgAnAMgn/171-home-page-creation-and-setup-migration-for-cms

#### What's this PR do?
fixes #596, Home page creation and setup migration for CMS

#### How should this be manually tested?
Just run the migration and check the home page exists

#### Screenshots
<img width="1440" alt="Screenshot 2020-06-11 at 13 28 15" src="https://user-images.githubusercontent.com/4043989/84362971-7a8bd180-abe7-11ea-9b76-726a5065011d.png">

ahmed-belal commented on .vscode/launch.json:
```diff
@@ -0,0 +1,21 @@
+      ]
+    }
+  ]
+}
```
Please remove this file from version control

umarmughal824 replied:
already removed.

ahmed-belal commented on cms/migrations/0017_setup_home_page.py:
```diff
@@ -0,0 +1,62 @@
+
+from django.db import migrations
+
+DEFAULT_HOMEPAGE_PROPS = dict(title="Home Page", subhead="This is the home page")
```
The default title should be "MIT Bootcamps", subhead should be "Learn Innovation and Entrepreneurship from MIT"

ahmed-belal replied:
Also, it should be `tagline`, not `subhead`.

ahmed-belal commented on cms/migrations/0017_setup_home_page.py:
```diff
@@ -0,0 +1,62 @@
+    if site is None:
+        Site.objects.create(
+            is_default_site=True, root_page=valid_home_page, **DEFAULT_SITE_PROPS
+        )
```
This should be checked earlier than `valid_home_page`

ahmed-belal commented on cms/migrations/0017_setup_home_page.py:
```diff
@@ -0,0 +1,62 @@
+        depth=2, content_type=ContentType.objects.get_for_model(Page)
+    ).first()
+    if wagtail_default_home_page is not None:
+        wagtail_default_home_page.delete()
```
You need to move all the existing children of the default wagtail home page under the new home page otherwise this migration will erase all page content from the CMS.

ahmed-belal commented on cms/migrations/0017_create_and_setup_home_page.py:
```diff
@@ -0,0 +1,60 @@
+        valid_home_page = HomePage(**DEFAULT_HOMEPAGE_PROPS)
+        root.add_child(instance=valid_home_page)
+        valid_home_page.refresh_from_db()
+    if site is None:
+        Site.objects.create(
+            is_default_site=True, root_page=valid_home_page, **DEFAULT_SITE_PROPS
+        )
```
This should be done before you try to fetch `valid_home_page`.

umarmughal824 replied:
but I can't create the site object without any root page so it requires that home page that's why it coming here.

ahmed-belal commented on cms/migrations/0017_create_and_setup_home_page.py:
```diff
@@ -0,0 +1,60 @@
+    wagtail_default_home_page = Page.objects.filter(
+        depth=2, content_type=ContentType.objects.get_for_model(Page)
+    ).first()
+    if wagtail_default_home_page is not None:
+        wagtail_default_home_page.delete()
```
You need to transfer any existing child page under the new page before you delete the default home page otherwise this migration will empty out the CMS.

umarmughal824 replied:
Yes that can be done becuase that's true it make the cms empty and home page starts to give `404`.

ahmed-belal commented on cms/migrations/0017_create_and_setup_home_page.py:
```diff
@@ -0,0 +1,60 @@
+        content_type=ContentType.objects.get_for_model(HomePage)
+    ).first()
+    home_page_descendants = home_page.get_descendants()
+    for descendant in home_page_descendants:
+        descendant.delete()
+    home_page.delete()
```
This is incorrect. You need to create a new default wagtail home page with title "Welcome to your new Wagtail site!" and also transfer any existing home page descendants under the default home page before you delete the one created earlier by applying this migration. Point being, if I apply and then rollback this migration my CMS and database should remain the same.

ahmed-belal commented on cms/migrations/0017_create_and_setup_home_page.py:
```diff
@@ -0,0 +1,60 @@
+        Site.objects.create(
+            is_default_site=True, root_page=valid_home_page, **DEFAULT_SITE_PROPS
+        )
+    elif site.root_page is None or site.root_page != valid_home_page:
+        site.root_page = valid_home_page
+        site.save()
+    wagtail_default_home_page = Page.objects.filter(
+        depth=2, content_type=ContentType.objects.get_for_model(Page)
+    ).first()
```
You can check `site.root_page` for type. If it's a `Page` object with "Welcome to your new Wagtail site!" title this is the default home page.

umarmughal824 replied:
this is not needed. Already addressed your feedback. everything should be in the flow now.

umarmughal824 replied:
It should work as it is now.

umarmughal824 commented:
@ahmed-belal I have addressed your feedback
- On applying migration, the default wagtail home page gets deleted and all of its child pages moved under the HomePage as child
- On un-applying migration, HomePage gets deleted and created a new default page and moved all the home page child pages under it

gsidebo commented on cms/migrations/0017_create_and_setup_home_page.py:
```diff
@@ -0,0 +1,74 @@
+    """
+    Create the home page
+    """
+    site = Site.objects.filter(is_default_site=True).first()
```
This can be optimized a bit

```python
from django.conf import settings

def create_home_page(apps, schema_editor):
    # ... get model classes from 'apps' ...
    root = Page.objects.get(depth=1)
    valid_home_page = Page.objects.filter(
        content_type=ContentType.objects.get_for_model(HomePage)
    ).first()
    if valid_home_page is None:
        valid_home_page = HomePage(**DEFAULT_HOMEPAGE_PROPS)
        root.add_child(instance=valid_home_page)
        valid_home_page.refresh_from_db()
    site, _ = Site.objects.get_or_create(
        is_default_site=True,
        defaults=dict(
            root_page=valid_home_page,
            site_name=settings.WAGTAIL_SITE_NAME,
            **DEFAULT_SITE_PROPS
        )
    )
    if site.root_page != valid_home_page:
        site.root_page = valid_home_page
        site.save()
    wagtail_default_home_page = Page.objects.filter(
        depth=2, content_type=ContentType.objects.get_for_model(Page)
    ).first()
    # ...
```

umarmughal824 commented on cms/migrations/0017_create_and_setup_home_page.py:
```diff
@@ -0,0 +1,74 @@
+from django.contrib.contenttypes.models import ContentType
+from wagtail.core.models import Page, Site
+
+from cms.models import HomePage
```
I have done it the same way already but it does not load most of the functions. like `add_child`. If you see I have 15 commits in history that are for the same reason to show the reviewer.

umarmughal824 replied:
See the build was failing as long as I was importing the models from `apps`.

umarmughal824 replied:
Please see the commits older than it.
```try with direct model imports.  75ad48c```
The build started to pass only after import direct because most of the functions were not available.

umarmughal824 replied:
I was getting very strange type of errors like

- `ValueError: Cannot query "ContentType object (3)": Must be "ContentType" instance.`
- `AttributeError: 'Page' object has no attribute 'add_child`
- `AttributeError: 'StateApps' object has no attribute 'get_mode'`

umarmughal824 replied:
Even I have verified it once more by importing models locally from `apps`.
```
umerasghar@A006-00274  ~/arbisoft/MIT/bootcamp-ecommerce   umar/bootcamp-596 ●  docker-compose exec web python manage.py migrate cms 0016_admissions_section
Operations to perform:
  Target specific migration: 0016_admissions_section, from cms
Running migrations:
  Rendering model states... DONE
  Unapplying cms.0017_create_and_setup_home_page...Traceback (most recent call last):
  File "manage.py", line 14, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 323, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 364, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 83, in wrapped
    res = handle_func(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/commands/migrate.py", line 234, in handle
    fake_initial=fake_initial,
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 121, in migrate
    state = self._migrate_all_backwards(plan, full_plan, fake=fake)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 196, in _migrate_all_backwards
    self.unapply_migration(states[migration], migration, fake=fake)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 269, in unapply_migration
    state = migration.unapply(state, schema_editor)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/migration.py", line 175, in unapply
    operation.database_backwards(self.app_label, schema_editor, from_state, to_state)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/operations/special.py", line 196, in database_backwards
    self.reverse_code(from_state.apps, schema_editor)
  File "/src/cms/migrations/0017_create_and_setup_home_page.py", line 66, in remove_home_page
    root.add_child(instance=default_wagtail_home_page)
AttributeError: 'Page' object has no attribute 'add_child'
 ✘ umerasghar@A006-00274  ~/arbisoft/MIT/bootcamp-ecommerce   umar/bootcamp-596 ● 
```

umarmughal824 replied:
Here is the old commit reference importing models from `apps` in the create and remove home page
https://github.com/mitodl/bootcamp-ecommerce/pull/631/commits/450adafdd8166dfeb4d0677d7f99706859801ec3#diff-3013b1b1421ce96a635d182d299283fcR14

asadiqbal08 replied:
@umarmughal824 as per my knowledge, usually in migrations, we prefer to use `apps.get_model(...)` because it uses the historical versions of the model and will not complain if some field(s) in future do not exist in code base and sure then migration can complain.

umarmughal824 replied:
yes I know that's the preferred way but we are not getting the required functions that way.

gsidebo replied:
I just got it to work this way:

```python
def create_home_page(apps, schema_editor):
    """
    Create the home page
    """
    HomePage = apps.get_model("cms", "HomePage")

    root = Page.objects.get(depth=1)
    home_page_content_type = ContentType.objects.get_for_model(HomePage)
    valid_home_page = Page.objects.filter(
        content_type_id=home_page_content_type.id
    ).first()
    if valid_home_page is None:
        valid_home_page = HomePage(
            **DEFAULT_HOMEPAGE_PROPS,
            content_type_id=home_page_content_type.id
        )
        root.add_child(instance=valid_home_page)
        valid_home_page = Page.objects.filter(
            content_type_id=home_page_content_type.id
        ).get()
    site, _ = Site.objects.get_or_create(
        is_default_site=True,
        defaults=dict(
            root_page_id=valid_home_page.id,
            site_name=settings.WAGTAIL_SITE_NAME,
            **DEFAULT_SITE_PROPS,
        ),
    )
    if site.root_page_id != valid_home_page.id:
        site.root_page_id = valid_home_page.id
        site.save()
    wagtail_default_home_page = Page.objects.filter(
        depth=2, content_type=ContentType.objects.get_for_model(Page)
    ).first()
    # ...
```

It's a bit more code, but I would really prefer that we at least load our own models (`HomePage`) via `apps.get_model`. You can load `HomePage` in the delete function in the same way with no other changes to that function

umarmughal824 commented on cms/migrations/0017_create_and_setup_home_page.py:
```diff
@@ -0,0 +1,74 @@
+        depth=2, content_type=ContentType.objects.get_for_model(Page)
+    ).first()
+    if wagtail_default_home_page is not None:
+        descendants = wagtail_default_home_page.get_descendants()
```
@ahmed-belal could you please confirm as you mentioned to use only descendants?

asadiqbal08 replied:
I have the same concerns about `get_descendants`.

umarmughal824 commented:
@gsidebo I have made a little change in the `create_home_page` because of getting that error.
<img width="1440" alt="Screenshot 2020-06-16 at 19 42 24" src="https://user-images.githubusercontent.com/4043989/84790569-ec5e8380-b00a-11ea-98c4-0b987cb84ce8.png">

umarmughal824 commented:
@gsidebo But I have noticed one issue which is that cms pages are not more accessible.
<img width="1404" alt="Screenshot 2020-06-16 at 20 31 06" src="https://user-images.githubusercontent.com/4043989/84795147-8bd24500-b010-11ea-92f0-a25fd1ac90be.png">
<img width="1440" alt="Screenshot 2020-06-16 at 20 31 18" src="https://user-images.githubusercontent.com/4043989/84795226-a2789c00-b010-11ea-8408-84767e3650de.png">

Are you getting that error?

umarmughal824 commented:
I have tested it the error is being produced by `get_children` but if I try it with `get_descendants`, it works fine.
<img width="1402" alt="Screenshot 2020-06-16 at 20 39 54" src="https://user-images.githubusercontent.com/4043989/84795973-904b2d80-b011-11ea-9cfd-70a9c18e0c39.png">
